### PR TITLE
Fix for signature malleability

### PR DESCRIPTION
**Pull request by yrong**
Resolves: https://linear.app/snowfork/issue/SNO-1215

### CODE REVIEW

File: relayer/relays/beefy/parameters.go [hunk 1]
```diff
@@ -89,13 +93,37 @@ func toBeefyClientCommitment(c *types.Commitment) *contracts.BeefyClientCommitme
 	}
 }
 
-func cleanSignature(input types.BeefySignature) (uint8, [32]byte, [32]byte) {
+func cleanSignature(input types.BeefySignature) (v uint8, r [32]byte, s [32]byte, err error) {
```
Comment (claravanstaden) [inline relayer/relays/beefy/parameters.go:96]: Out of interest, why name the return parameters?

Comment (yrong) [reply relayer/relays/beefy/parameters.go:96]: Nothing special, just for return with error https://github.com/Snowfork/snowbridge/blob/aa41d7d32f0955317f2201336c8b78040d3c76e9/relayer/relays/beefy/parameters.go#L106 easily.

File: relayer/cmd/root.go [hunk 2]
```diff
@@ -22,9 +22,9 @@ var rootCmd = &cobra.Command{
 func init() {
 	rootCmd.AddCommand(run.Command())
 	rootCmd.AddCommand(getBlockCmd())
-	//rootCmd.AddCommand(fetchMessagesCmd())
```
Comment (alistair-singh) [inline relayer/cmd/root.go:25]: Commented code?

File: relayer/relays/beefy/parameters.go [hunk 3]
```diff
@@ -89,13 +93,42 @@ func toBeefyClientCommitment(c *types.Commitment) *contracts.BeefyClientCommitme
 	}
 }
 
-func cleanSignature(input types.BeefySignature) (uint8, [32]byte, [32]byte) {
+func CleanSignature(input types.BeefySignature) (v uint8, r [32]byte, s [32]byte, reverted bool, err error) {
```
Comment (alistair-singh) [inline relayer/relays/beefy/parameters.go:96]: Is there any link to documentation about this algorithm that we can add here? Maybe a link to the polkadot implementation?

**Review by alistair-singh — APPROVED** [review]

File: relayer/relays/beefy/parameters.go [hunk 4]
```diff
@@ -89,13 +93,42 @@ func toBeefyClientCommitment(c *types.Commitment) *contracts.BeefyClientCommitme
...
+	s = *(*[32]byte)(input[32:64])
+	v = uint8(input[64])
+	if v < 27 {
+		v += 27
+	}
+	if v != 27 && v != 28 {
+		return v, r, s, reverted, fmt.Errorf("invalid V:%d", v)
+	}
+	var N *uint256.Int = uint256.NewInt(0)
+	N.SetFromHex("0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141")
```
Comment (alistair-singh) [inline relayer/relays/beefy/parameters.go:110]: Nit: Use upper case for consistency.
```suggestion
	N.SetFromHex("0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141")
```

File: relayer/relays/beefy/parameters.go [hunk 5]
```diff
@@ -89,13 +93,43 @@ func toBeefyClientCommitment(c *types.Commitment) *contracts.BeefyClientCommitme
...
+		negativeS256 = negativeS256.Sub(N, s256)
+		s = negativeS256.Bytes32()
+		if v%2 == 0 {
+			v = v - 1
+		} else {
+			v = v + 1
+		}
+		reverted = true
+	}
+	return v, r, s, reverted, nil
```
Comment (vgeddes) [inline relayer/relays/beefy/parameters.go:132]: The `reverted` variable doesn't seem to be used anywhere?

Comment (yrong) [reply relayer/relays/beefy/parameters.go:132]: https://github.com/Snowfork/snowbridge/blob/ce46a00c5fb83d759d9e4dba119f72bd8e2570c9/relayer/cmd/scan_beefy.go#L183

Just used to inspect the beefy signatures.

**Review by vgeddes — APPROVED** [review]

**Pull request merged**